up-l1/README.adoc

```diff
@@ -50,157 +42,218 @@ We will use the following terminology during this document:
-
-
-NOTE: uProtocol data model used by the uTransport interface are declared in https://github.com/eclipse-uprotocol/up-core-api[up-core-api]
+NOTE: The data types used in the following sections are defined in link:../basics/README.adoc[uProtocol Basic Types].
 
 == UListener
 
-`UListener` is an interface used for receiving messages by the client uE from the uTransport layer. Listeners are used for receiving messages asynchronously by the uTransport layer.  
+A uEntity registers a `UListener` with the transport in order to process (incoming) messages that are of interest to the uEntity.
 
-  * Type name *MUST* be `UListener`
-  * *MUST* contain the method `void onReceive(UMessage&)` that is called by the Transport to notify (callback) the client
+=== OnReceive
 
+A transport invokes this method for each newly arrived message that matches the criteria specified as part of <<_registerlistener,registering the listener>>.
 
-== UTransport
+[source]
+----
+onReceive(IN message: UMessage&)
+----
 
-The transport layer API responsible for point-2-point communication. The interface provides common functionality across all transport implementations.
+==== Parameters
+.OnReceive Parameters
+[width="100%",cols="15%,15%,70%"]
+|===
+|Parameter | Type | Description
 
-* *MUST* implement all APIs defined in this section
+| message
+| link:../basics/umessage.adoc[UMessage]
+| The newly received message.
 
-=== Send()
+|===
 
-Transmit a UMessage.  
 
+== UTransport
 
-Communication protocols (a.k.a. uProtocol Transports) define their own Protocol Data Unit (PDU) that comprises of header and payload. Some transports header parameters map already to uProtocol UAttributes. Sending UAttributes in the transport's payload as well as in the header increases overhead for little to no gain. To address this issue,  <<Send()>> allows flexibility for uTransport implementers to define however they see fit what gets mapped into their transport header vs payload. 
+This is the main entry point into a transport's messaging functionality.
 
-For example, if HartleyTransport can map only UPriority to its PDU header, then HartleyTransport would define a PDU payload type that includes UPayload and all the other UAttributes and send that new type in its PDU payload such that the receiver will not loose any metadata. 
+=== Send
 
-The API signature:
+Clients use this method to transmit a single message.
 
-`OUT UStatus send(UMessage&)`
+[source]
+----
+send(
+  IN message: UMessage&,
+  OUT result: Future<Void, UStatus>
+)
+----
 
-==== Parameters
 .Send Parameters
-[width="100%",cols="20%,80%"]
+[width="100%",cols="15%,25%,60%"]
 |===
-|Parameter | Description
-
-| UMessage
-
-| UProtocol message containing the header (UAttributes) and payload (UPayload)
+|Parameter | Type | Description
 
- * *MUST* be passed by reference
+| message
+| link:../basics/umessage.adoc[UMessage]
+| The message to send.
 
+| result
+| Future<Void, link:../basics/error_model.adoc[UStatus]>
```

Comment:
I wonder why I would get back a UStatus instead of Void if sending the message succeeded? FMPOV this would result in a pretty clunky developer experience which would require me to always explicitly check for the UCode to be 0, or am I mistaken?

Reply:
Clunky yes but simpler than checking if it is void or UStatus etc...

Reply:
IMHO this
```java
try {
  transport.send(myMessage);
} catch (SendException e) {
  log.error(e);
}
```

is much preferable over this
```java
var result = transport.send(myMessage);
if (result.isError()) {
  log.error("failed to send message: {}", result.getCode());
} else {
  // proceed
}
```
This is also true for Rust and I guess every other language that has an exception mechanism. Using that programming language's exception mechanism seems very desirable to me ...

Reply:
Perhaps a bit of my Rust bias creeping in here, but we have language-level mechanisms to make discriminating success-vs-failure cases in `Result<T, E>`.

### Modeled with explicit success and failure at the type level:
```rust
fn register_listener(
  sourceFilter: UUri,
  sinkFilter: UUri?,
  listener: UListener&
) -> Result<(), UStatus>;
```
allows us to easily:
```rust
let registration_result = client.register_listener(source_filter, sink_filter, listener);

// use pattern matching
match registration_result {
  Err(ustatus) => { /* handle error case with access to `code` and `message` */ }
  Ok(()) => { /* handle success */ }
}

// use if-let pattern
if let Err(ustatus) = registration_result {
  /* handle just the error condition inside of ustatus */
}
```

### Modeled without explicit success and failure at the type level:
```rust
fn register_listener(
  sourceFilter: UUri,
  sinkFilter: UUri?,
  listener: UListener&
) -> UStatus;
```
we then need to:
```rust
let registration_result = client.register_listener(source_filter, sink_filter, listener);

// unpack the UCode and message to then match on and use both
let (code, message) = (registration_result.get_code(), registration_result.get_message());
match code {
  UCode::OK => { /* handle the success case */ }
   _ => { /* handle the remaining cases of `code` and also use `message` */ }
}

// unable to use if-let, must fall back to an unpack and if condition
let (code, message) = (registration_result.get_code(), registration_result.get_message());
if code != UCode::OK {
  /* handle the cases where not a success using `code` and `message` */
}
```

Could we ask for other language implementors to weigh in here with what this would look like?